code/obj/item/material.dm

```diff
@@ -29,15 +30,45 @@
 		setup_material()
 		if(src.material?.getName())
 			initial_material_name = src.material.getName()
+		_update_stack_appearance()
 
 	proc/setup_material() // Overwrite for ore specific setup
 		return
 
 	_update_stack_appearance()
-		if(material)
-			UpdateName(src) // get the name in order so it has whatever it needs
-			name = "[amount] [src.name][amount > 1 ? "s":""]"
-		return
+		if(!material)
+			return
+		update_stack_name()
+		var/icon_stack_new = get_stack_value()
+		if(icon_stack_new && icon_stack_new != src.icon_stack_value) // Only update icon_state if it needs updating (0 to ignore)
+			src.icon_stack_value = icon_stack_new
+			UpdateIcon()
+
+	update_icon()
+		src.icon_state = "ore[src.icon_stack_value]_$$[src.material.getName()]"
+
+	proc/update_stack_name() // How the material should be named at different stack sizes
+		UpdateName(src) // get the name in order so it has whatever it needs
+		if(src.amount == 1)
+			name = "[src.name]"
+		else
+			name = "[amount] [src.name]s"
+
+	proc/get_stack_value() // Determines at what stack sizes the icon_state changes
+		// example: src.icon_state = "ore[x]_$$gold"
+		switch(src.amount)
+			if(1)
+				return 1
+			if(2 to 9)
```

Comment:
Could you scrunch these ranges down a little? Most material stacks are going to be in the 1-10 range and it seems wrong to only have two icon states for that.